project/src/backend/sdl/SDLGamepad.cpp

```diff
@@ -2,90 +2,104 @@
 
 
 namespace lime {
+	std::map<int, SDLGamepad> gameControllers;
+	std::map<int, int> gameControllerIDs;
+	
+	void SDLGamepad::Rumble(int duration, double largeStrength, double smallStrength) {
+		// Make sure game controller is open
+		if (gameController == nullptr)
+			return;
+		
+		// Rumble controller
+		if (smallStrength < 0.0f)
+			smallStrength = 0.0f;
+		else if (smallStrength > 1.0f)
+			smallStrength = 1.0f;
+		
+		if (largeStrength < 0.0f)
+			largeStrength = 0.0f;
+		else if (largeStrength > 1.0f)
+			largeStrength = 1.0f;
+		
+		if (duration < 0)
+			duration = 0;
+		else if (duration > 0xFFFF)
+			duration = 0xFFFF;
+		
+		SDL_GameControllerRumble(gameController, largeStrength * 0xFFFF, smallStrength * 0xFFFF, duration);
+	}
 
-
-	std::map<int, SDL_GameController*> gameControllers = std::map<int, SDL_GameController*> ();
-	std::map<int, int> gameControllerIDs = std::map<int, int> ();
-
-
+	// SDL static gamepad API
 	bool SDLGamepad::Connect (int deviceID) {
-
 		if (SDL_IsGameController (deviceID)) {
+			SDL_GameController *gameController = SDL_GameControllerOpen(deviceID);
+			
+			if (gameController != nullptr) {
+				SDL_Joystick *joystick = SDL_GameControllerGetJoystick(gameController);
+				int id = SDL_JoystickInstanceID(joystick);
 
-			SDL_GameController *gameController = SDL_GameControllerOpen (deviceID);
-
-			if (gameController) {
-
-				SDL_Joystick *joystick = SDL_GameControllerGetJoystick (gameController);
-				int id = SDL_JoystickInstanceID (joystick);
-
-				gameControllers[id] = gameController;
+				gameControllers[id] = std::move(SDLGamepad(gameController));
 				gameControllerIDs[deviceID] = id;
-
 				return true;
-
 			}
-
 		}
-
 		return false;
 
 	}
 
 
 	bool SDLGamepad::Disconnect (int id) {
-
-		if (gameControllers.find (id) != gameControllers.end ()) {
-
-			SDL_GameController *gameController = gameControllers[id];
-			SDL_GameControllerClose (gameController);
-			gameControllers.erase (id);
-
+		if (gameControllers.find(id) != gameControllers.end()) {
+			gameControllers.erase(id);
 			return true;
-
 		}
-
 		return false;
 
 	}
 
 
 	int SDLGamepad::GetInstanceID (int deviceID) {
-
 		return gameControllerIDs[deviceID];
-
 	}
 
 
+	// Gamepad API
 	void Gamepad::AddMapping (const char* content) {
-
 		SDL_GameControllerAddMapping (content);
-
 	}
 
 
 	const char* Gamepad::GetDeviceGUID (int id) {
-
-		SDL_Joystick* joystick = SDL_GameControllerGetJoystick (gameControllers[id]);
+		auto it = gameControllers.find(id);
+		if (it == gameControllers.end())
+			return nullptr;
+		
+		SDL_Joystick* joystick = SDL_GameControllerGetJoystick (it->second.gameController);
 
 		if (joystick) {
-
 			char* guid = new char[64];
 			SDL_JoystickGetGUIDString (SDL_JoystickGetGUID (joystick), guid, 64);
 			return guid;
-
 		}
-
-		return 0;
+		
```

Comment:
Try to follow a project's existing formatting convention, even if you don't like it. Leave the newlines, strip the tabs, etc.

(I personally do like having tabs on otherwise blank lines, but only if done consistently.)

Reply:
i did try to keep the formatting consistent, however I didn't notice the tabs as that's handled automatically by my IDE.

Reply:
I don't always notice them either, but fortunately GitHub highlights whitespace changes.